pkg/metadata/manager.go

```diff
@@ -0,0 +1,374 @@
+	for {
+		select {
+		case ev := <-m.podEvents:
+			if !m.enabled {
+				continue
+			}
+			switch ev.eventType {
+			case PodAddEvent, PodUpdateEvent:
+				m.updatePodMetadata(ev.pod, ev.oldPod)
+			case PodDeleteEvent:
+				m.deletePodMetadata(ev.pod)
+			case PodMetadataEvent:
+				m.getParentPodMetadata(ev.cgroupId, ev.podMetadataOut)
+			}
+		case <-m.shutdown:
+			return
+		}
+	}
+}
+
+func (m *CgroupManager) updatePodMetadata(pod, oldPod *v1.Pod) {
+	id := string(pod.ObjectMeta.UID)
+	pm, ok := m.podMetadataById[id]
+	if !ok {
+		// Fill in pod static metadata.
+		pm = &podMetadata{
+			// TODO: remove id?
+			id:        id,
+			name:      pod.Name,
+			namespace: pod.Namespace,
+		}
+		m.podMetadataById[id] = pm
+	}
+	if oldPod != nil && oldPod.Status.DeepEqual(&pod.Status) {
+		return
+	}
+	// Only update the metadata that can change. This excludes pod's name,
+	// namespace, id, and qos class.
+	podIPs := pod.Status.PodIPs
+	pm.ips = make([]string, len(podIPs))
+	for i := range podIPs {
+		pm.ips[i] = podIPs[i].IP
+	}
+	// Get metadata for pod's containers that are in the running state. Containers
+	// can get re-created, and their ids can change. Update the new containers.
+	// Pod's metadata including its containers map will be deleted when the pod
+	// is deleted.
+	numContainers := len(pod.Status.ContainerStatuses)
+	if pm.containers == nil && numContainers > 0 {
+		pm.containers = make(map[string]struct{})
+	}
+	currContainers := make(map[string]struct{}, numContainers)
+	for _, c := range pod.Status.ContainerStatuses {
+		var cId string
+		if cId = c.ContainerID; cId == "" || c.State.Running == nil {
+			continue
+		}
+		// The container ID field is of the form: <container-runtime>://<containerID>
+		// Example:containerd://e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0e
+		_, after, found := strings.Cut(cId, "//")
+		if !found || after == "" {
+			log.Errorf("unexpected container ID: %s", cId)
+			continue
+		}
+		cId = after
+		if _, ok := pm.containers[cId]; ok {
+			currContainers[cId] = struct{}{}
+			// Container cgroup path doesn't change as long as the container id
+			// is the same.
+			continue
+		}
+		pm.containers[cId] = struct{}{}
+		currContainers[cId] = struct{}{}
+
+		// Container could've been gone, so don't log any errors.
+		cgrpPath, err := m.getContainerCgroupPath(id, cId, pod.Status.QOSClass)
+		if err != nil {
+			log.Debugf("failed to get container metadata for (%s): %v", cId, err)
+			continue
+		}
+		cgrpId, err := m.cgroupsChecker.GetCgroupID(cgrpPath)
+		if err != nil {
+			log.Debugf("failed to get cgroup id for cgroup path (%s): %v", cgrpPath, err)
+			continue
+		}
+		log.Infof("aditi-debug-metadata %s %d", cgrpPath, cgrpId)
+		m.containerMetadataByCgrpId[cgrpId] = &containerMetadata{
+			cgroupId:    cgrpId,
+			cgroupPath:  cgrpPath,
+			parentPodId: id,
+		}
+	}
+	// Clean up any pod's old containers.
+	if oldPod != nil {
+		for _, c := range oldPod.Status.ContainerStatuses {
```

Comment:
why not `range pm.containers` here?

Reply:
Well, we first update `pm.containers` based on the latest state, and then clean up from the old pod containers. Didn't seem worthwhile to first do a diff.